### ISSUE TITLE
[v93k] Need a way to handle volatile flags

### ISSUE THREAD

**Issue by arkohler** [Origen-SDK/origen_testers#55]
In some cases there are flags that are changed by the test method.  The current flow generator will optimize away multiple checks to the same flag in the same state.  There should be a way to override this behavior for cases where a flag needs to be treated as volatile.
In the code below, assume test4 can change the value of flag1.
```
  log 'Nested flag optimiztion test case'
  if_flag :flag1 do
    func :test4, id: :nf_t4
    if_failed :nf_t4 do
      render 'multi_bin;', if_flag: :flag1
    end
  end
```
Generates this output, without a second check of @FLAG1.
```
  print_dl("Nested flag optimization test case");
  if @FLAG1 == 1 then
  {
    run_and_branch(test4_6_BEA7F3B)
    then
    {
    }
    else
    {
      multi_bin;
    }
  }
  else
  {
  }

```

**Comment by ginty** [comment]
So I'm not sure if your use case is common enough that we should change the default behavior, but that is an open question and changing the current behavior is certainly one of the options on the table.
Here are the main options I can see:

- Consider any explicit references to a flag via `if/unless_flag` as potentially being volatile. Origen then should not perform any optimizations on these and only user flow variables (`if/unless_enable`) and Origen-generated flags would be optimized away where possible.
- As above, but only consider the flag volatile if it is not otherwise set by Origen APIs elsewhere in the current flow. i.e. if the flag looks to be owned by a 3rd party and not the current Origen flow.
- Provide an API for the application code to specify which flag(s) should be treated as volatile.

Any thoughts on the way to go @arkohler.  @rlaj, @redxeth, any thoughts?

**Comment by arkohler** [comment]
I agree that it should be relatively rare.  In our case, the test methods will set an Alarm flag in the case of supply alarms.  It's then up to the flow to check the flag and bin accordingly.  In the case of nested tests, each test will need to check this flag.  This leads to the test case shown above.

Of the three options above, my vote would go to the third option.  Changing the default behavior for a rare case, I think, is probably overkill.  Perhaps something like:
`  if_flag :flag1, :volatile true do
`
A volatile: option could be used to force the optimizer to ignore it so that the flag is placed in the flow as given in the ruby file.
It might be a bit more tricky in the single line case since the it might be harder to tell which options are for the parent vs the flag.
`render 'multi_bin;', if_flag: :flag1, volatile: true`
Although it may not be an issue since, a volatile: option would only have meaning to a flag.

I'm interested to head feedback from other users.

**Comment by redxeth** [comment]
Hmmm..  changed my mind-- I was thinking #3 but after thoughts as to implementation I figured that if you go and say you want a flag then that means you know what you're doing and don't want Origen to optimize or clean it up for you.   I'm thinking #1 now.

Currently am thinking a different flag-related thing-- setting the initial state in the var file to non-zero-- still looking to see if the API supports this.

**Comment by ginty** [comment]
Not really a fan of #1 tbh, I can imagine being back here with someone complaining that their flag references are not generating pretty V93K flow logic in future.

For the volatile API, I would propose defining which flags should be treated as volatile at the start of the flow like this:

~~~ruby
Flow.create do
  
  volatile "$Alarm", "$Timeout"

  # Tests that reference these flags here

  # The volatile definition would obviously apply to any imported sub-flows too
  import "some_sub_flow"

end
~~~

**Comment by redxeth** [comment]
So if you're defining which are volatile, then don't need the `$` right?   I'm ok with having volatile commands.

**Comment by info-rchitect** [comment]
I know these are flow flags, these volatile flags really sound like test events versus flags that check pass/fail.  i.e. if i define a test event like an Alarm, wouldn't I want to perhaps do something besides flow branching?  When I wrote test programs I used to purposely force a clamp Alarm, check for it, and then run some code to decipher which power supply set it.  These test events could be callbacks in the test interface rather than handled in flow files.  IMO this makes more sense because when a particular type of test event occurs, you want to do the same thing over and over.  Seems much DRYer to me.  Anyway, just some thoughts when considering an implementation.

**Comment by info-rchitect** [comment]
this issue has cooled a bit but unfortunately we need a solution very soon.  what is the general consensus here?  @ginty @arkohler @redxeth @rlaj

**Comment by chrisnappi** [comment]
Seems to me like everyone would be okay with Stephen's last suggestion?

**Comment by ginty** [comment]
Yeah I hope so, 'cause it's already half-implemented :-)

This PR adds the underlying infrastructure for this, needs to be integrated into OrigenTesters now, which shouldn't take too long: https://github.com/Origen-SDK/atp/pull/5
